### ISSUE TITLE
Handling of attributes containing js expressions

### ISSUE THREAD

**Issue by hhaensel** [GenieFramework/Stipple.jl#231]
In many cases vue uses js expressions, e.g. arrays of objects (dicts), to set options.

Currently we handle such situations mostly by defining a model field that takes a Julia expression which is then converted to the correct js expression during rendering. Sometimes, we also set js expressions explicitly, e.g. `options = R"['hello', 'world']"`.

I propose to modify Stipple attribute handling such that the latter example could be written as `options = ["hello", "world"]`. That kind of handling would also allow easier definitions of objects or array of objects:
```julia
julia> btntoggle(:network, options = [opts(label = x, value = lowercase(x)) for x in ("Google", "Facebook")])
"<q-btn-toggle v-model=\"network\" :options=\"[{'value':'google','label':'Google'},{'value':'facebook','label':'Facebook'}]\"></q-btn-toggle>"
```

**Comment by PGimenez** [comment]
Mmm I don't know about this. I see how it can be useful as you avoid having to define extra variables, but I don't think it's much simpler than doing

```julia
@out option_list = [opts(label = x, value = lowercase(x)) for x in ("Google", "Facebook")]
btntoggle(:network, options = :option_list)
```

Moreover, this wouldn't prevent the instantiation of the list, so I think there wouldn't be much computation time savings (not sure about memory).
 
Also, as of now, anything you pass to a parameter has to be a symbol or a Javascript expression. Same goes for macros like `@on`. If we extend them to accepting list comprehensions, couldn't this lead to some confusion?

**Comment by hhaensel** [comment]
> Moreover, this wouldn't prevent the instantiation of the list, so I think there wouldn't be much computation time savings (not sure about memory).

That's not true. The list is only instantiated at the client side and it will not consume memory per model, because the list definition will reside in Stipple.js. It's only defined once UI.
```julia
btntoggle(:network, options = [opts(label = x, value = lowercase(x)) for x in ("Google", "Facebook")])
```

> Also, as of now, anything you pass to a parameter has to be a symbol or a Javascript expression. Same goes for macros like `@on`. If we extend them to accepting list comprehensions, couldn't this lead to some confusion?

Currently, parameter values can be either
- `AbstractString`, `Real`: The value is directly passed to the parameter as a constant at the client side
- `Symbol`: The string value of the symbol is evaluated as a js expression and is updated if one of the included variables changes

So I think if we add
- `<julia expression, e.g. Dict, Vector, GenericTrace>`: The values are jsonified and evaluated as js expressions

that's only very intuitive, think e.g. of `select(:field, options = [1, 2, 3])`, `plot(:plot, layout = PlotlyBase.Layout(title = "test title"))`

**Comment by hhaensel** [comment]
> Same goes for macros like `@on`

The content of the `@on` or `@click` macro is executable code. We don't have any representation for that in Julia, so I don't see how you would want to change the macro to handle julia expressions.

I know there is a package with a macro that translates julia code to js code, but that's a different story. Although, I would find it quite nice for `js_methods()` to write julia code and have it executed as javascript on the client side. Let's think about that. I need to dig, which package that was.

**Comment by PGimenez** [comment]
I was also thinking of `@recur`. I've seen people attempt to put Julia expressions in there and ask why it doesn't work. With this change one would expect something like `@recur([fruit in fruit_list])`to work

**Comment by hhaensel** [comment]
Good point! Andfine for me and easy to implement.

**Comment by PGimenez** [comment]
Well the second way would probably be better as you get, the iterated variable. What I wrote would be akin to doing `for _ in fruit_list`. Still, I wrote it as a list comprehension because otherwise it'd have to be an Exp, right? 

Here's an example of what I meant from one of our users

<img width="531" alt="image" src="https://github.com/GenieFramework/Stipple.jl/assets/5058397/dda6c156-6e48-4b8b-acd0-1b81b435f362">

**Comment by hhaensel** [comment]
Well, as `@recur` is a macro, anything inside is an expression first of all. So we could do what we want with it.
Javascript enumeration works just the other way around, the index follows second. So, indeed we'd need that translator package.

**Comment by hhaensel** [comment]
There is JSExpr.jl which does a lotvof what we'd want. Only Dict translation is not done and there is  o zip function, but we could add that as method

**Comment by hhaensel** [comment]
> Here's an example of what I meant from one of our users
> 
> <img alt="image" width="531" src="https://user-images.githubusercontent.com/5058397/278296900-dda6c156-6e48-4b8b-acd0-1b81b435f362.png">

Have a look at https://vuejs.org/guide/essentials/list.html#v-for

You find that you can easily access the index by destructuring
```julia
@recur("for (letter, i) in letters")
```
where `i` will run from `0` to `length(letters) -1` and will be the correct js index so that `letter == letters[i]` for each element.

#### zip
If you have another vaiable that you want to loop, instead of 
```javascript
for (var1, var2) in zip(vars1, vars2)
```
you can do
```javascript
for (var1, i) in vars1
```
and write `vars2[i]` instead of `var2` in the referencing expression.

**Comment by hhaensel** [comment]
But we could implement Julia expressions as well. Question would be how exactly. We could either use expression evaluation or explicit interpolation

1. expression evaluation
```julia
julia> dict = Dict(:a => "b", :c => 4);

julia> list(@recur((v, k, i) in mydict), "{{ v }}")
"<q-list v-for=\"(v, k, i) in {'a':'b','c':4}\">{{ v }}</q-list>"

julia> list(@recur((v, k, i) in :mydict), "{{ v }}")
"<q-list v-for=\"(v, k, i) in mydict>{{ v }}</q-list>"
```
2: interpolation
```julia
list(@recur((v, k, i) in $mydict), "{{ v }}")
"<q-list v-for=\"(v, k, i) in {'a':'b','c':4}\">{{ v }}</q-list>"

list(@recur((v, k, i) in mydict), "{{ v }}")
"<q-list v-for=\"(v, k, i) in mydict>{{ v }}</q-list>"
```

**Comment by hhaensel** [comment]
Re-reading my proposals, I would vote for the first solution.

**Comment by hhaensel** [comment]
@PGimenez @essenciary 
If version 1) also meets your taste, I think we're set to merge this PR.

My idea is the following:
- merge #234 
- merge #232 

Prepare another PR to support julia expressions in `@for` (alias for '@recur`)

Are there other macros that would need to support Julia expressions?JSExpr is quite a heavy dependency and I'm not sure how much that would help, after I tried some examples. Please share your opinion.

**Comment by essenciary** [comment]
Yes, looks good to me!
I would not add JSExpr - users could add it themselves maybe.

**Comment by hhaensel** [comment]
Solved by #232
